CS3219-AY2223S1/cs3219-project-ay2223s1-g12
# Implement Login and Logout

Fixes #22 

Summary
* Implement JWT creation upon successful login
* Add login route `/api/user/login`
* ~~This version of authentication is in development before the merge of PR #8 , hence authentication involves comparison of plaintext passwords~~
* Implement password comparison using bcrypt,  Closes #24 
* Fix bcrypt import typo in services.js
* Update invalid login error messages to be more specific. Closes #28 
* Implement refresh of tokens after expiry
* Implement logout of token. Closes #29 
* Abstract orm functions into _user-orm.js_ Closes #27 
* Add check for user existence in orm. Closes #31 

Steps to test:
1. `npm install` in the `user-services` folder. This should install the jwt and bcrypt packages if not already installed.
2. Follow demo in project development guide to create a default demo user in db, if not already done so.
3. LOGIN: Make a POST API call via route `http://localhost:8000/api/user/login` with request body:
```
{
    "username": "cs3219",
    "password": "cs3219"
}
```
4. Expected positive result:
```
{
    "message": "Login Success!",
    "token": "xxxxxx"
    "refreshToken": "yyyyy"
}
```
5. Expected negative result:
```
{
    "message": "Authentication failed. Invalid username." // or password
}
```
6. To test AUTH USER: Make a GET to `http://localhost:8000/api/user/`, paste logged-in user's token value (from `/login` response) into request header's Authorization Value within 30s (current token validity time).
![image](https://user-images.githubusercontent.com/12153882/191728310-fcd7817d-3be6-4f7f-a25f-84549c5c2c6b.png)
or
![image](https://user-images.githubusercontent.com/12153882/191728418-8f4f1b08-631f-44be-ac8f-0b618ae8d65d.png)
Expected output:
```
{
    "message": "Authenticated {your username here}"
}
```

7. To test LOGOUT: Make a POST to `http://localhost:8000/api/user/logout`. In the request body:
```
{
    "refreshToken" : "{your user's refresh token here}" 
}
```
Expected result:
```
{
    "message": "Logout successful!"
}
```

Logout at this point of time works by revoking a user's refreshToken, which coupled with a short access token validity prevents them from refreshing their expired access token against the refreshToken to gain a new access token
